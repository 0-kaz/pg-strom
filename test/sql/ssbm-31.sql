SET search_path = pgstrom_regress,public;
SET pg_strom.debug_kernel_source = off;
--Q3_1
EXPLAIN(costs off, verbose)
select c_nation, s_nation, d_year, sum(lo_revenue)
as revenue from customer, lineorder, supplier, date1
where lo_custkey = c_custkey
and lo_suppkey = s_suppkey
and lo_orderdate = d_datekey
and c_region = 'ASIA'  and s_region = 'ASIA'
and d_year >= 1992 and d_year <= 1997
  group by c_nation, s_nation, d_year
  order by d_year asc, revenue desc;

select c_nation, s_nation, d_year, sum(lo_revenue)
as revenue from customer, lineorder, supplier, date1
where lo_custkey = c_custkey
and lo_suppkey = s_suppkey
and lo_orderdate = d_datekey
and c_region = 'ASIA'  and s_region = 'ASIA'
and d_year >= 1992 and d_year <= 1997
  group by c_nation, s_nation, d_year
  order by d_year asc, revenue desc;
