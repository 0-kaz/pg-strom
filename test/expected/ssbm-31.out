SET search_path = pgstrom_regress,public;
SET pg_strom.debug_kernel_source = off;
--Q3_1
EXPLAIN(costs off, verbose)
select c_nation, s_nation, d_year, sum(lo_revenue)
as revenue from customer, lineorder, supplier, date1
where lo_custkey = c_custkey
and lo_suppkey = s_suppkey
and lo_orderdate = d_datekey
and c_region = 'ASIA'  and s_region = 'ASIA'
and d_year >= 1992 and d_year <= 1997
  group by c_nation, s_nation, d_year
  order by d_year asc, revenue desc;
                                                                                 QUERY PLAN                                                                                 
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort
   Output: customer.c_nation, supplier.s_nation, date1.d_year, (pgstrom.fsum_numeric((pgstrom.psum((lineorder.lo_revenue)::double precision))))
   Sort Key: date1.d_year, (pgstrom.fsum_numeric((pgstrom.psum((lineorder.lo_revenue)::double precision)))) DESC
   ->  GroupAggregate
         Output: customer.c_nation, supplier.s_nation, date1.d_year, pgstrom.fsum_numeric((pgstrom.psum((lineorder.lo_revenue)::double precision)))
         Group Key: customer.c_nation, supplier.s_nation, date1.d_year
         ->  Sort
               Output: customer.c_nation, supplier.s_nation, date1.d_year, (pgstrom.psum((lineorder.lo_revenue)::double precision))
               Sort Key: customer.c_nation, supplier.s_nation, date1.d_year
               ->  Gather
                     Output: customer.c_nation, supplier.s_nation, date1.d_year, (pgstrom.psum((lineorder.lo_revenue)::double precision))
                     Workers Planned: 2
                     ->  Parallel Custom Scan (GpuPreAgg)
                           Output: customer.c_nation, supplier.s_nation, date1.d_year, (pgstrom.psum((lineorder.lo_revenue)::double precision))
                           Reduction: Local
                           GPU Projection: customer.c_nation, supplier.s_nation, date1.d_year, lineorder.lo_revenue, pgstrom.psum((lineorder.lo_revenue)::double precision)
                           Combined GpuJoin: enabled
                           ->  Parallel Custom Scan (GpuJoin) on pgstrom_regress.lineorder
                                 Output: customer.c_nation, supplier.s_nation, date1.d_year, lineorder.lo_revenue
                                 GPU Projection: customer.c_nation::character(15), supplier.s_nation::character(15), date1.d_year::integer, lineorder.lo_revenue::numeric
                                 Outer Scan: pgstrom_regress.lineorder
                                 Depth 1: GpuHashJoin
                                          HashKeys: lineorder.lo_suppkey
                                          JoinQuals: (lineorder.lo_suppkey = supplier.s_suppkey)
                                 Depth 2: GpuHashJoin
                                          HashKeys: lineorder.lo_custkey
                                          JoinQuals: (lineorder.lo_custkey = customer.c_custkey)
                                 Depth 3: GpuHashJoin
                                          HashKeys: lineorder.lo_orderdate
                                          JoinQuals: (lineorder.lo_orderdate = date1.d_datekey)
                                 ->  Seq Scan on pgstrom_regress.supplier
                                       Output: supplier.s_nation, supplier.s_suppkey
                                       Filter: (supplier.s_region = 'ASIA'::bpchar)
                                 ->  Seq Scan on pgstrom_regress.customer
                                       Output: customer.c_nation, customer.c_custkey
                                       Filter: (customer.c_region = 'ASIA'::bpchar)
                                 ->  Seq Scan on pgstrom_regress.date1
                                       Output: date1.d_year, date1.d_datekey
                                       Filter: ((date1.d_year >= 1992) AND (date1.d_year <= 1997))
(39 rows)

select c_nation, s_nation, d_year, sum(lo_revenue)
as revenue from customer, lineorder, supplier, date1
where lo_custkey = c_custkey
and lo_suppkey = s_suppkey
and lo_orderdate = d_datekey
and c_region = 'ASIA'  and s_region = 'ASIA'
and d_year >= 1992 and d_year <= 1997
  group by c_nation, s_nation, d_year
  order by d_year asc, revenue desc;
    c_nation     |    s_nation     | d_year |   revenue   
-----------------+-----------------+--------+-------------
 VIETNAM         | VIETNAM         |   1992 | 33573660759
 INDONESIA       | VIETNAM         |   1992 | 21231485964
 CHINA           | VIETNAM         |   1992 | 21000236620
 CHINA           | CHINA           |   1992 | 19627209204
 CHINA           | INDIA           |   1992 | 15068912776
 INDIA           | INDONESIA       |   1992 | 14484860871
 JAPAN           | INDONESIA       |   1992 | 14239197074
 CHINA           | JAPAN           |   1992 | 14203350787
 JAPAN           | INDIA           |   1992 | 13766110705
 INDONESIA       | CHINA           |   1992 | 13700489631
 INDIA           | VIETNAM         |   1992 | 12882361859
 JAPAN           | VIETNAM         |   1992 | 12643685946
 CHINA           | INDONESIA       |   1992 |  7851642217
 INDIA           | JAPAN           |   1992 |  7733004490
 VIETNAM         | INDIA           |   1992 |  7656652088
 VIETNAM         | JAPAN           |   1992 |  7526136671
 INDONESIA       | INDIA           |   1992 |  6982951677
 JAPAN           | CHINA           |   1992 |  6872681471
 INDONESIA       | JAPAN           |   1992 |  6854539284
 INDIA           | CHINA           |   1992 |  6402742804
 INDONESIA       | INDONESIA       |   1993 | 36538856359
 CHINA           | INDONESIA       |   1993 | 21997221121
 CHINA           | JAPAN           |   1993 | 21458814603
 JAPAN           | VIETNAM         |   1993 | 20269705306
 VIETNAM         | VIETNAM         |   1993 | 16458386848
 JAPAN           | INDONESIA       |   1993 | 15539510152
 JAPAN           | INDIA           |   1993 | 14704452377
 INDONESIA       | VIETNAM         |   1993 | 14639691440
 VIETNAM         | CHINA           |   1993 | 14495716052
 INDIA           | INDONESIA       |   1993 | 13948607782
 INDIA           | VIETNAM         |   1993 | 13363650800
 INDIA           | CHINA           |   1993 | 13328913166
 INDONESIA       | INDIA           |   1993 |  8086573834
 VIETNAM         | INDIA           |   1993 |  7683790784
 CHINA           | INDIA           |   1993 |  7618324026
 INDONESIA       | JAPAN           |   1993 |  6910441618
 VIETNAM         | INDONESIA       |   1993 |  6548025881
 CHINA           | VIETNAM         |   1993 |  6538131683
 INDIA           | JAPAN           |   1993 |  6243657347
 INDONESIA       | INDONESIA       |   1994 | 36133944592
 JAPAN           | INDONESIA       |   1994 | 21866384696
 INDONESIA       | VIETNAM         |   1994 | 21308088912
 CHINA           | JAPAN           |   1994 | 20811437322
 INDIA           | VIETNAM         |   1994 | 20392902376
 JAPAN           | JAPAN           |   1994 | 17534417325
 CHINA           | INDONESIA       |   1994 | 15337218804
 INDONESIA       | INDIA           |   1994 | 15075547426
 VIETNAM         | CHINA           |   1994 | 14805000243
 VIETNAM         | JAPAN           |   1994 | 14013469477
 INDIA           | JAPAN           |   1994 | 13652501920
 INDIA           | CHINA           |   1994 | 13335166948
 CHINA           | INDIA           |   1994 |  8298766393
 JAPAN           | INDIA           |   1994 |  7960976735
 INDIA           | INDONESIA       |   1994 |  7021173136
 INDONESIA       | CHINA           |   1994 |  6754745612
 JAPAN           | VIETNAM         |   1994 |  6660103307
 CHINA           | VIETNAM         |   1994 |  6268029073
 CHINA           | INDONESIA       |   1995 | 21926413993
 INDIA           | JAPAN           |   1995 | 21119589581
 INDONESIA       | INDONESIA       |   1995 | 19743444843
 VIETNAM         | VIETNAM         |   1995 | 16731181096
 JAPAN           | JAPAN           |   1995 | 16659973292
 INDONESIA       | INDIA           |   1995 | 15524532894
 INDONESIA       | VIETNAM         |   1995 | 14882009437
 CHINA           | INDIA           |   1995 | 14705564863
 INDIA           | VIETNAM         |   1995 | 14680139059
 JAPAN           | INDONESIA       |   1995 | 14428333690
 VIETNAM         | JAPAN           |   1995 | 14321765363
 CHINA           | VIETNAM         |   1995 | 14267801226
 JAPAN           | CHINA           |   1995 | 13617443878
 CHINA           | JAPAN           |   1995 |  8167534118
 INDIA           | CHINA           |   1995 |  7029680342
 INDONESIA       | JAPAN           |   1995 |  6989017212
 INDIA           | INDONESIA       |   1995 |  6850607170
 JAPAN           | VIETNAM         |   1995 |  6743794260
 VIETNAM         | INDONESIA       |   1995 |  6630772812
 VIETNAM         | CHINA           |   1995 |  6437168877
 VIETNAM         | INDIA           |   1995 |  6412776763
 CHINA           | CHINA           |   1996 | 36028870070
 INDONESIA       | JAPAN           |   1996 | 22303907710
 INDONESIA       | INDIA           |   1996 | 21765012074
 CHINA           | INDIA           |   1996 | 21457977737
 VIETNAM         | INDONESIA       |   1996 | 21171121079
 INDIA           | INDIA           |   1996 | 16718878609
 CHINA           | INDONESIA       |   1996 | 15520038788
 VIETNAM         | INDIA           |   1996 | 14498967309
 CHINA           | JAPAN           |   1996 | 14321395987
 JAPAN           | INDIA           |   1996 | 13816469844
 CHINA           | VIETNAM         |   1996 | 13566285089
 JAPAN           | VIETNAM         |   1996 | 13524703833
 VIETNAM         | CHINA           |   1996 |  7474090277
 VIETNAM         | JAPAN           |   1996 |  7437856733
 JAPAN           | CHINA           |   1996 |  7073828854
 INDONESIA       | CHINA           |   1996 |  7031166581
 INDIA           | JAPAN           |   1996 |  6758321981
 INDIA           | VIETNAM         |   1996 |  6699217366
 JAPAN           | INDONESIA       |   1997 | 22179757437
 JAPAN           | JAPAN           |   1997 | 19567616936
 INDIA           | INDIA           |   1997 | 17099398028
 VIETNAM         | VIETNAM         |   1997 | 15960392452
 INDIA           | INDONESIA       |   1997 | 15736605969
 VIETNAM         | INDONESIA       |   1997 | 14916910836
 INDIA           | JAPAN           |   1997 | 14698772855
 CHINA           | JAPAN           |   1997 | 14531036128
 JAPAN           | VIETNAM         |   1997 | 14286616412
 VIETNAM         | CHINA           |   1997 | 14238161962
 INDONESIA       | CHINA           |   1997 | 14215376015
 CHINA           | INDIA           |   1997 | 13601276570
 INDIA           | VIETNAM         |   1997 | 13129416291
 CHINA           | INDONESIA       |   1997 |  7915704554
 INDIA           | CHINA           |   1997 |  7814088029
 VIETNAM         | INDIA           |   1997 |  7393083368
 INDONESIA       | VIETNAM         |   1997 |  6729544463
 JAPAN           | CHINA           |   1997 |  6642459887
 INDONESIA       | INDIA           |   1997 |  6583500531
 CHINA           | VIETNAM         |   1997 |  6529129303
 JAPAN           | INDIA           |   1997 |  6492173025
 VIETNAM         | JAPAN           |   1997 |  6147161575
(118 rows)

